product_default_image/models/product_template.py

```diff
@@ -0,0 +1,100 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 LasLabs Inc.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+from odoo import api, fields, models, tools
+
+
+class ProductTemplate(models.Model):
+
+    _inherit = 'product.template'
+
+    categ_change_image = fields.Boolean(
+        help='Used to prevent automatic changes to image fields when categ_id '
+        'changes after products have been created. '
+        '(False prevents change)',
+    )
+
+    @api.model
+    def default_get(self, vals):
+        vals = super(ProductTemplate, self).default_get(vals)
+
+        if not vals.get('categ_id'):
+            return vals
+
+        company = self.env.user.company_id
+        category_mod = self.env['product.category']
+        target = company.product_image_target
+
+        if target == 'category':
+            categ = category_mod.browse(vals['categ_id'])
+            vals['categ_change_image'] = True
+            vals['image'] = categ.image
+            vals['image_medium'] = categ.image_medium
+            vals['image_small'] = categ.image_small
+
+        elif target == 'global':
+            vals['image'] = company.product_default_global_image
+            tools.image_resize_images(vals)
+
+        return vals
+
+    @api.multi
+    @api.onchange('categ_id')
+    def _onchange_categ_id(self):
+        """ Changes image to new categ_id's image.
+
+        This method checks if a custom image has been uploaded. If not,
+        the image is changed to the new categ_id's image.
+
+        Cases where it is assumed a custom image has NOT been uploaded:
+            * image, image_medium, and image_small are all empty
+            * image matches any category's image field
+            * image_medium matches any category's image_medium field
+            * image_small matches any category's image_small field
+
+        image, image_medium, and image_small fields are all compared against
+        categories because, for example, in certain cases the image field
+        is None, but the image_medium field has data.
+        """
+        for record in self:
+
+            if not self.env.user.company_id.product_image_target == 'category':
```

Comment:
This isn't using the record, so should be evaluated before the loop. It would result in the loop simply iterating through itself if the condition is true.